_LEARNING-CIRCLE-24 PROJECT_

<h2> Learning-Circle-24 </h2> <p>  Contributors </p>



# **Crime in Chicago**
This is a dataset
The data dictionary can be found in [Chicago Crime Portal](https://data.cityofchicago.org/Public-Safety/Crimes-2001-to-Present/ijzp-q8t2/about_data)

In this notebook, we are going to explore more about crime in Chicago and try to answer few questions.

Analyzing the Chicago crime dataset involves examining various aspects of the data to understand patterns, trends, and characteristics of criminal activities in Chicago. We perform analysis using Python, pandas and visualization libraries using the following steps:

1. Load the Dataset: Load the Chicago crime dataset into a DataFrame.
2. Explore the Dataset: Examine the structure, columns, and basic statistics of the dataset.
3. Data Preprocessing: We cleaned the data by dropping redundant columns,handling missing values and converting columns to their appropriate format.
4. Feature Engineering: Create new column for Months, Day, Season
5. EDA: Exploratory Data Analysis
    - What is the total number of crime cases filed from 2001 till date?
    - How many unique crimes are available and what is the count for each crime from 2001-2022?
    - what is the number of crime per year?
    - What is the number of crime per month?
    - Top 10 number of cases discovered each day?
    - What is the trend of crime by type?
    - Where does crimes happen the most?
    - What is the number of crime discovered by month and year?
    - Are there specific location that crimes used to happen?
    - What is the arrested rate per crime?
    - Top 10 Block with the most crime?
    - Top 10 ward with most amount of crimes?
    - Top 10 community area with the most amount of crimes?
    - Top 10 communities with the most crime in 2023?
    - Top 10 Location where theft normally occurs?
    - Top-15 crime where arrest occured?
    - How has crime in Chicago changed across years? Was 2023 really the bloodiest year in two decades?
    - Are some types of crimes more likely to happen in specific locations or specific time of the day or specific day of the week than other types of        crimes?
    - In which month did the lowest motor vehicle thefts occur?
    - How prominant are certain types of various crimes?
    - How do rates of crime change of time? For instance, are there any seasonal effects or changes depending on the time ofday?
    - Where do these crimes occur? Are some areas more safe than others?
6. Summarize and Visualize Insights: Summarize and Visualize the key insights and findings from the analysis using plots and charts.

In [5]:
# importing required libraries
import pandas as pd
import numpy as np
import seaborn as sns
import plotly.express as px
import matplotlib.pyplot as plt
%matplotlib inline
from datetime import datetime

In [6]:
# Set display option
pd.set_option('display.max.columns', None)

# 1. Loading the dataset

This dataset was extracted in 2023 and contains crime cases from 2001 to 2023, but we want the data for only cases from 2001 to 2022 due to incomplete cases for 2023 in order to avoid a bias analysis.

In [7]:
full_dataset = pd.read_csv('crime_data_chicago.csv')
# Drop rows that have 2024 as the year.
chicago_crime_df = pd.DataFrame(full_dataset[full_dataset['Year'] != 2023])
chicago_crime_df.head()

Unnamed: 0        ID Case Number                    Date  \
0           0   6407111    HP485721  07/26/2008 02:30:00 PM   
1           1  11398199    JB372830  07/31/2018 10:57:00 AM   
2           2   5488785    HN308568  04/27/2007 10:30:00 AM   
3           3  11389116    JB361368  07/23/2018 08:55:00 AM   
4           4  12420431    JE297624  07/11/2021 06:40:00 AM   

                  Block  IUCR       Primary Type               Description  \
0  085XX S MUSKEGON AVE  1320    CRIMINAL DAMAGE                TO VEHICLE   
1     092XX S ELLIS AVE  143C  WEAPONS VIOLATION  UNLAWFUL POSS AMMUNITION   
2     062XX N TRIPP AVE  0610           BURGLARY            FORCIBLE ENTRY   
3    0000X N KEELER AVE  0560            ASSAULT                    SIMPLE   
4   016XX W HARRISON ST  051A            ASSAULT      AGGRAVATED - HANDGUN   

                     Location Description  Arrest  Domestic  Beat  District  \
0                                  STREET   False     False   423       4.0   
1                               POOL ROOM    True     False   413       4.0   
2                               RESIDENCE    True     False  1711      17.0   
3            NURSING HOME/RETIREMENT HOME   False     False  1115      11.0   
4  PARKING LOT / GARAGE (NON RESIDENTIAL)   False     False  1231      12.0   

   Ward  Community Area FBI Code  X Coordinate  Y Coordinate  Year  \
0  10.0            46.0       14     1196638.0     1848800.0  2008   
1   8.0            47.0       15     1184499.0     1843935.0  2018   
2  39.0            12.0       05     1146911.0     1941022.0  2007   
3  28.0            26.0      08A     1148388.0     1899882.0  2018   
4  27.0            28.0      04A     1165430.0     1897441.0  2021   

               Updated On   Latitude  Longitude                       Location  
0  02/28/2018 03:56:25 PM  41.739980 -87.555120  (41.739979622, -87.555120042)  
1  08/07/2018 04:02:59 PM  41.726922 -87.599747  (41.726922145, -87.599746995)  
2  02/28/2018 03:56:25 PM  41.994138 -87.734959  (41.994137622, -87.734959049)  
3  07/30/2018 03:52:24 PM  41.881217 -87.730590  (41.881217483, -87.730589961)  
4  07/18/2021 04:56:02 PM  41.874174 -87.668082  (41.874173691, -87.668082118)

# 2. Explore the dataset

In [8]:
#checking the dimension of the dataset
chicago_crime_df.shape

(2278726, 23)

In [9]:
# Display the first five rows of the dataset
chicago_crime_df.columns

Index(['Unnamed: 0', 'ID', 'Case Number', 'Date', 'Block', 'IUCR',
       'Primary Type', 'Description', 'Location Description', 'Arrest',
       'Domestic', 'Beat', 'District', 'Ward', 'Community Area', 'FBI Code',
       'X Coordinate', 'Y Coordinate', 'Year', 'Updated On', 'Latitude',
       'Longitude', 'Location'],
      dtype='object')

In [10]:
# check the first five observations
chicago_crime_df.head()

Unnamed: 0        ID Case Number                    Date  \
0           0   6407111    HP485721  07/26/2008 02:30:00 PM   
1           1  11398199    JB372830  07/31/2018 10:57:00 AM   
2           2   5488785    HN308568  04/27/2007 10:30:00 AM   
3           3  11389116    JB361368  07/23/2018 08:55:00 AM   
4           4  12420431    JE297624  07/11/2021 06:40:00 AM   

                  Block  IUCR       Primary Type               Description  \
0  085XX S MUSKEGON AVE  1320    CRIMINAL DAMAGE                TO VEHICLE   
1     092XX S ELLIS AVE  143C  WEAPONS VIOLATION  UNLAWFUL POSS AMMUNITION   
2     062XX N TRIPP AVE  0610           BURGLARY            FORCIBLE ENTRY   
3    0000X N KEELER AVE  0560            ASSAULT                    SIMPLE   
4   016XX W HARRISON ST  051A            ASSAULT      AGGRAVATED - HANDGUN   

                     Location Description  Arrest  Domestic  Beat  District  \
0                                  STREET   False     False   423       4.0   
1                               POOL ROOM    True     False   413       4.0   
2                               RESIDENCE    True     False  1711      17.0   
3            NURSING HOME/RETIREMENT HOME   False     False  1115      11.0   
4  PARKING LOT / GARAGE (NON RESIDENTIAL)   False     False  1231      12.0   

   Ward  Community Area FBI Code  X Coordinate  Y Coordinate  Year  \
0  10.0            46.0       14     1196638.0     1848800.0  2008   
1   8.0            47.0       15     1184499.0     1843935.0  2018   
2  39.0            12.0       05     1146911.0     1941022.0  2007   
3  28.0            26.0      08A     1148388.0     1899882.0  2018   
4  27.0            28.0      04A     1165430.0     1897441.0  2021   

               Updated On   Latitude  Longitude                       Location  
0  02/28/2018 03:56:25 PM  41.739980 -87.555120  (41.739979622, -87.555120042)  
1  08/07/2018 04:02:59 PM  41.726922 -87.599747  (41.726922145, -87.599746995)  
2  02/28/2018 03:56:25 PM  41.994138 -87.734959  (41.994137622, -87.734959049)  
3  07/30/2018 03:52:24 PM  41.881217 -87.730590  (41.881217483, -87.730589961)  
4  07/18/2021 04:56:02 PM  41.874174 -87.668082  (41.874173691, -87.668082118)

In [11]:
 # Display column information and data types
chicago_crime_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2278726 entries, 0 to 2278725
Data columns (total 23 columns):
 #   Column                Dtype  
---  ------                -----  
 0   Unnamed: 0            int64  
 1   ID                    int64  
 2   Case Number           object 
 3   Date                  object 
 4   Block                 object 
 5   IUCR                  object 
 6   Primary Type          object 
 7   Description           object 
 8   Location Description  object 
 9   Arrest                bool   
 10  Domestic              bool   
 11  Beat                  int64  
 12  District              float64
 13  Ward                  float64
 14  Community Area        float64
 15  FBI Code              object 
 16  X Coordinate          float64
 17  Y Coordinate          float64
 18  Year                  int64  
 19  Updated On            object 
 20  Latitude              float64
 21  Longitude             float64
 22  Location              object 
dtypes: bool

In [12]:
# checking missing value count
chicago_crime_df.isnull().sum()

Unnamed: 0                   0
ID                           0
Case Number                  1
Date                         0
Block                        0
IUCR                         0
Primary Type                 0
Description                  0
Location Description      2877
Arrest                       0
Domestic                     0
Beat                         0
District                    12
Ward                    184695
Community Area          184267
FBI Code                     0
X Coordinate             23985
Y Coordinate             23985
Year                         0
Updated On                   0
Latitude                 23985
Longitude                23985
Location                 23985
dtype: int64

In [13]:
chicago_crime_df.describe()

Unnamed: 0            ID          Beat      District          Ward  \
count  2.278726e+06  2.278726e+06  2.278726e+06  2.278714e+06  2.094031e+06   
mean   1.139362e+06  6.882068e+06  1.186442e+03  1.129072e+01  2.272764e+01   
std    6.578117e+05  3.419168e+06  7.026836e+02  6.946692e+00  1.383464e+01   
min    0.000000e+00  6.370000e+02  1.110000e+02  1.000000e+00  1.000000e+00   
25%    5.696812e+05  3.716076e+06  6.210000e+02  6.000000e+00  1.000000e+01   
50%    1.139362e+06  6.885990e+06  1.034000e+03  1.000000e+01  2.300000e+01   
75%    1.709044e+06  9.887568e+06  1.731000e+03  1.700000e+01  3.400000e+01   
max    2.278725e+06  1.278199e+07  2.535000e+03  3.100000e+01  5.000000e+01   

       Community Area  X Coordinate  Y Coordinate          Year      Latitude  \
count    2.094459e+06  2.254741e+06  2.254741e+06  2.278726e+06  2.254741e+06   
mean     3.752140e+01  1.164569e+06  1.885747e+06  2.009638e+03  4.184209e+01   
std      2.153282e+01  1.673955e+04  3.209855e+04  6.019724e+00  8.830434e-02   
min      0.000000e+00  0.000000e+00  0.000000e+00  2.001000e+03  3.661945e+01   
25%      2.300000e+01  1.152948e+06  1.859053e+06  2.004000e+03  4.176866e+01   
50%      3.200000e+01  1.166060e+06  1.890673e+06  2.009000e+03  4.185578e+01   
75%      5.700000e+01  1.176365e+06  1.909219e+06  2.014000e+03  4.190668e+01   
max      7.700000e+01  1.205119e+06  1.951622e+06  2.022000e+03  4.202291e+01   

          Longitude  
count  2.254741e+06  
mean  -8.767161e+01  
std    6.073538e-02  
min   -9.168657e+01  
25%   -8.771379e+01  
50%   -8.766597e+01  
75%   -8.762823e+01  
max   -8.752453e+01

# 3. Data Preprocessing

**Data Cleaning**

Here are some steps that we would carry out in this phase

1. Remove redundant or unnecessary columns
2. Handle missing values
3. Convert data types to the appropriate formats  

__Reasons for dropping the below columns__
- 'Unnamed: 0': Its just an index for individual row/ Observation
- 'ID': Serial identifier for the record which is redundant due to the presence of a case number for each case.
- 'Location': This is where the incident occurred in a format that allows for creation of maps and other geographic operation. It's the combination of both latitude and longitude feature of this dataset thus, not required.
- 'Updated On': Holds the date and time the record was last updated.Which is not required for our analysis.
- 'X Coordinate': The x coordinate is a shifted location from the actual location for partial redaction but falls on the same block. Since we have a longitude/latitude and the block where the crime happened, we can neglect this feature.
- 'Y Coordinate': The y coordinate is a shifted location from the actual location for partial redaction but falls on the same block. Since we have a longitude/latitude and the block where the crime happened, we can neglect this feature.
- 'Community Area': This indicates the community area where the incident occurred. Which is not neccessary for our analysis.
- 'IUCR': This is the Illinois Unifrom Crime Reporting code. This is directly linked to the Primary Type and Description, it provides information about the primary type and description of crime. Since primary type and description features already exist, it becomes a redundant column
- Description: It's simply a subcategory of the primary type(which describes the type of crime committed). this can be omitted
- Block: Since we have the latitude and longitude, this column is redundant.
- Community Area: This is not needed in our analysis and the presence of a police Beat makes this redundant.
- FBI Code: Removed because it is not necessary for our analysis as we are unable to de-code these codes.

__1. Use a funtion to Drop redundant columns and/or columns that are not necessary for our analysis (a redundant column is one that gives infomation that another column already explains: eg ID and Case number)__

In [14]:
chicago_crime_df.columns

Index(['Unnamed: 0', 'ID', 'Case Number', 'Date', 'Block', 'IUCR',
       'Primary Type', 'Description', 'Location Description', 'Arrest',
       'Domestic', 'Beat', 'District', 'Ward', 'Community Area', 'FBI Code',
       'X Coordinate', 'Y Coordinate', 'Year', 'Updated On', 'Latitude',
       'Longitude', 'Location'],
      dtype='object')

In [15]:
# Dropping redundant columns
def drop_redundant_columns(df):
    redundant_columns = ['Unnamed: 0','ID', 'Location','Updated On', 'FBI Code', 'Description', 'X Coordinate','Y Coordinate','IUCR']
    df.drop(redundant_columns, axis=1, inplace= True)

drop_redundant_columns(chicago_crime_df)

In [16]:
chicago_crime_df.head()

Case Number                    Date                 Block  \
0    HP485721  07/26/2008 02:30:00 PM  085XX S MUSKEGON AVE   
1    JB372830  07/31/2018 10:57:00 AM     092XX S ELLIS AVE   
2    HN308568  04/27/2007 10:30:00 AM     062XX N TRIPP AVE   
3    JB361368  07/23/2018 08:55:00 AM    0000X N KEELER AVE   
4    JE297624  07/11/2021 06:40:00 AM   016XX W HARRISON ST   

        Primary Type                    Location Description  Arrest  \
0    CRIMINAL DAMAGE                                  STREET   False   
1  WEAPONS VIOLATION                               POOL ROOM    True   
2           BURGLARY                               RESIDENCE    True   
3            ASSAULT            NURSING HOME/RETIREMENT HOME   False   
4            ASSAULT  PARKING LOT / GARAGE (NON RESIDENTIAL)   False   

   Domestic  Beat  District  Ward  Community Area  Year   Latitude  Longitude  
0     False   423       4.0  10.0            46.0  2008  41.739980 -87.555120  
1     False   413       4.0   8.0            47.0  2018  41.726922 -87.599747  
2     False  1711      17.0  39.0            12.0  2007  41.994138 -87.734959  
3     False  1115      11.0  28.0            26.0  2018  41.881217 -87.730590  
4     False  1231      12.0  27.0            28.0  2021  41.874174 -87.668082

In [17]:
chicago_crime_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2278726 entries, 0 to 2278725
Data columns (total 14 columns):
 #   Column                Dtype  
---  ------                -----  
 0   Case Number           object 
 1   Date                  object 
 2   Block                 object 
 3   Primary Type          object 
 4   Location Description  object 
 5   Arrest                bool   
 6   Domestic              bool   
 7   Beat                  int64  
 8   District              float64
 9   Ward                  float64
 10  Community Area        float64
 11  Year                  int64  
 12  Latitude              float64
 13  Longitude             float64
dtypes: bool(2), float64(5), int64(2), object(5)
memory usage: 213.0+ MB


In [18]:
# A summary statistics of each column
chicago_crime_df.describe(include='all')

Case Number                    Date             Block Primary Type  \
count      2278725                 2278726           2278726      2278726   
unique     2278678                 1343546             54347           35   
top       HK172551  01/01/2008 12:01:00 AM  100XX W OHARE ST        THEFT   
freq             3                      81              4878       479375   
mean           NaN                     NaN               NaN          NaN   
std            NaN                     NaN               NaN          NaN   
min            NaN                     NaN               NaN          NaN   
25%            NaN                     NaN               NaN          NaN   
50%            NaN                     NaN               NaN          NaN   
75%            NaN                     NaN               NaN          NaN   
max            NaN                     NaN               NaN          NaN   

       Location Description   Arrest Domestic          Beat      District  \
count               2275849  2278726  2278726  2.278726e+06  2.278714e+06   
unique                  198        2        2           NaN           NaN   
top                  STREET    False    False           NaN           NaN   
freq                 590957  1675252  1965311           NaN           NaN   
mean                    NaN      NaN      NaN  1.186442e+03  1.129072e+01   
std                     NaN      NaN      NaN  7.026836e+02  6.946692e+00   
min                     NaN      NaN      NaN  1.110000e+02  1.000000e+00   
25%                     NaN      NaN      NaN  6.210000e+02  6.000000e+00   
50%                     NaN      NaN      NaN  1.034000e+03  1.000000e+01   
75%                     NaN      NaN      NaN  1.731000e+03  1.700000e+01   
max                     NaN      NaN      NaN  2.535000e+03  3.100000e+01   

                Ward  Community Area          Year      Latitude     Longitude  
count   2.094031e+06    2.094459e+06  2.278726e+06  2.254741e+06  2.254741e+06  
unique           NaN             NaN           NaN           NaN           NaN  
top              NaN             NaN           NaN           NaN           NaN  
freq             NaN             NaN           NaN           NaN           NaN  
mean    2.272764e+01    3.752140e+01  2.009638e+03  4.184209e+01 -8.767161e+01  
std     1.383464e+01    2.153282e+01  6.019724e+00  8.830434e-02  6.073538e-02  
min     1.000000e+00    0.000000e+00  2.001000e+03  3.661945e+01 -9.168657e+01  
25%     1.000000e+01    2.300000e+01  2.004000e+03  4.176866e+01 -8.771379e+01  
50%     2.300000e+01    3.200000e+01  2.009000e+03  4.185578e+01 -8.766597e+01  
75%     3.400000e+01    5.700000e+01  2.014000e+03  4.190668e+01 -8.762823e+01  
max     5.000000e+01    7.700000e+01  2.022000e+03  4.202291e+01 -8.752453e+01

In [19]:
chicago_crime_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2278726 entries, 0 to 2278725
Data columns (total 14 columns):
 #   Column                Dtype  
---  ------                -----  
 0   Case Number           object 
 1   Date                  object 
 2   Block                 object 
 3   Primary Type          object 
 4   Location Description  object 
 5   Arrest                bool   
 6   Domestic              bool   
 7   Beat                  int64  
 8   District              float64
 9   Ward                  float64
 10  Community Area        float64
 11  Year                  int64  
 12  Latitude              float64
 13  Longitude             float64
dtypes: bool(2), float64(5), int64(2), object(5)
memory usage: 213.0+ MB


__3. Handling missing values__

In [20]:
# checking missing value count
chicago_crime_df.isnull().sum()

Case Number                  1
Date                         0
Block                        0
Primary Type                 0
Location Description      2877
Arrest                       0
Domestic                     0
Beat                         0
District                    12
Ward                    184695
Community Area          184267
Year                         0
Latitude                 23985
Longitude                23985
dtype: int64

In [21]:
# # Missing Values

# Function to calculate missing values by column
def missing_values_table(df):
        # Total missing values
        mis_val = df.isnull().sum()

        # Percentage of missing values
        mis_val_percent = 100 * df.isnull().sum() / len(df)

        # Make a table with the results
        mis_val_table = pd.concat([mis_val, mis_val_percent], axis=1)

        # Rename the columns
        mis_val_table_ren_columns = mis_val_table.rename(
        columns = {0 : 'Missing Values', 1 : '% of Total Values'})

        # Sort the table by percentage of missing descending
        mis_val_table_ren_columns = mis_val_table_ren_columns[
            mis_val_table_ren_columns.iloc[:,1] != 0].sort_values(
        '% of Total Values', ascending=False).round(1)

        # Print some summary information
        print ("Your selected dataframe has " + str(df.shape[1]) + " columns.\n"
            "There are " + str(mis_val_table_ren_columns.shape[0]) +
              " columns that have missing values.")

        # Return the dataframe with missing information
        return mis_val_table_ren_columns

In [22]:
missing_values_table(chicago_crime_df)

Your selected dataframe has 14 columns.
There are 7 columns that have missing values.


Missing Values  % of Total Values
Ward                          184695                8.1
Community Area                184267                8.1
Latitude                       23985                1.1
Longitude                      23985                1.1
Location Description            2877                0.1
District                          12                0.0
Case Number                        1                0.0

- __Handling Missing Case Number__

We have one (1) missing Case Number in the entire dataset. We have decided to drop the row where Case Number is missing

In [23]:
# dropping the missing Case Number column null value
chicago_crime_df.dropna(subset=['Case Number'],inplace=True) # i chose to drop this because the percentage NA is insignificant

In [24]:
chicago_crime_df.isna().sum()

Case Number                  0
Date                         0
Block                        0
Primary Type                 0
Location Description      2877
Arrest                       0
Domestic                     0
Beat                         0
District                    12
Ward                    184695
Community Area          184267
Year                         0
Latitude                 23984
Longitude                23984
dtype: int64

- __Handling Missing values for the District Column__

We have just 12 missing value for the district column which is 0.0% of the entire dataset. We have decided to drop the row where location discription is missing.

In [25]:
# dropping the missing district column null value
chicago_crime_df.dropna(subset=['District'],inplace=True) # i chose to drop this because the percentage NA is insignificant

In [26]:
chicago_crime_df.District.unique()

array([ 4., 17., 11., 12.,  2., 15., 25., 14.,  8.,  6., 19.,  1.,  7.,
       20., 18.,  3.,  9., 22.,  5., 16., 10., 24., 31.])

In [27]:
chicago_crime_df.isna().sum()

Case Number                  0
Date                         0
Block                        0
Primary Type                 0
Location Description      2877
Arrest                       0
Domestic                     0
Beat                         0
District                     0
Ward                    184695
Community Area          184267
Year                         0
Latitude                 23984
Longitude                23984
dtype: int64

- __Handling Missing values for the ward Column (Using data imputation)__

In [28]:
missing_values_table(chicago_crime_df)

Your selected dataframe has 14 columns.
There are 5 columns that have missing values.


Missing Values  % of Total Values
Ward                          184695                8.1
Community Area                184267                8.1
Latitude                       23984                1.1
Longitude                      23984                1.1
Location Description            2877                0.1

In [29]:
# checking the unique values of the ward column
chicago_crime_df.Ward.unique()

array([10.,  8., 39., 28., 27., nan, 29., 37., 26., 35., 14.,  1., 17.,
        3., 44., 36.,  4.,  2.,  6.,  7., 40., 42., 20., 13., 11., 34.,
       47., 33.,  9., 21., 45., 25., 24., 15., 16., 19., 43., 30., 23.,
       32., 31., 22., 50., 49.,  5., 46., 12., 41., 38., 18., 48.])

In [30]:
chicago_crime_df.Ward

0          10.0
1           8.0
2          39.0
3          28.0
4          27.0
           ... 
2278721    42.0
2278722     NaN
2278723     4.0
2278724    42.0
2278725     1.0
Name: Ward, Length: 2278713, dtype: float64

In [31]:
# Using ffill(forward fill) to fill missing value in the ward column
chicago_crime_df.Ward = chicago_crime_df.Ward.ffill() # Last observed value is carried forward

In [32]:
chicago_crime_df.isna().sum()

Case Number                  0
Date                         0
Block                        0
Primary Type                 0
Location Description      2877
Arrest                       0
Domestic                     0
Beat                         0
District                     0
Ward                         0
Community Area          184267
Year                         0
Latitude                 23984
Longitude                23984
dtype: int64

- __Handling Missing values for the Community Area Column (Using data imputation)__

In [33]:
missing_values_table(chicago_crime_df)

Your selected dataframe has 14 columns.
There are 4 columns that have missing values.


Missing Values  % of Total Values
Community Area                184267                8.1
Latitude                       23984                1.1
Longitude                      23984                1.1
Location Description            2877                0.1

In [34]:
# checking the unique values of the ward column
chicago_crime_df['Community Area'].unique()

array([46., 47., 12., 26., 28., nan, 25., 23., 22., 27., 63., 24., 71.,
       38.,  6., 19., 55., 36., 32., 69., 43.,  4.,  8., 42., 65., 60.,
       14., 49., 35., 40., 21., 73., 52., 15., 18., 16., 61., 66., 44.,
       45., 67., 74.,  2., 29., 51., 57., 17.,  7., 62., 30., 31., 13.,
       75.,  5.,  1., 39., 53.,  3., 68., 11., 37., 41., 48., 33., 59.,
       58., 76., 54., 10., 20., 34., 77., 64., 56., 70.,  9., 72., 50.,
        0.])

In [35]:
# Using ffill(forward fill) to fill missing value in the Community Area column
chicago_crime_df['Community Area'] = chicago_crime_df['Community Area'].ffill() # Last observed value is carried forward

In [36]:
chicago_crime_df.isna().sum()

Case Number                 0
Date                        0
Block                       0
Primary Type                0
Location Description     2877
Arrest                      0
Domestic                    0
Beat                        0
District                    0
Ward                        0
Community Area              0
Year                        0
Latitude                23984
Longitude               23984
dtype: int64

- __Handling Missing values for the latitude and Longitude Column__ (Using Assosciation) <br>
    - The assumption here is that the latitude and longitude readings are points that show exact location, are closer to eachother within a district. if they are, we can group our data by districts and fill up missing Latitide and longitude values by the median or mode.

In [37]:
chicago_crime_df.groupby(['District', 'Longitude', 'Latitude']).size()

District  Longitude   Latitude 
1.0       -87.776059  41.933944    1
          -87.645624  41.881909    4
          -87.645320  41.885582    2
          -87.645301  41.878079    1
          -87.645121  41.885586    1
                                  ..
31.0      -87.673838  41.776755    1
          -87.631942  41.930460    1
          -87.617227  41.646187    1
          -87.610757  41.884107    1
          -87.562704  41.737240    1
Length: 548193, dtype: int64

There appears to be closeness in longitude and Latitude points when data is grouped based on districts, we can group based on district and fill up by taking the median.

In [38]:
chicago_crime_df['Latitude'] = chicago_crime_df.groupby('District')['Latitude'].transform(lambda x: x.fillna(x.median()))
chicago_crime_df['Longitude'] = chicago_crime_df.groupby('District')['Longitude'].transform(lambda x: x.fillna(x.median()))

In [39]:
chicago_crime_df.isna().sum()

Case Number                0
Date                       0
Block                      0
Primary Type               0
Location Description    2877
Arrest                     0
Domestic                   0
Beat                       0
District                   0
Ward                       0
Community Area             0
Year                       0
Latitude                   0
Longitude                  0
dtype: int64

- __Handling Missing values for the Location Description__

We have just 2877 missing value for the location description which is 0.1% of the entire dataset. We can decide to drop the row where location decription is missing or fill this row with the most crime location.

In [40]:
chicago_crime_df.groupby(['District','Location Description']).size()

District  Location Description                       
1.0       ABANDONED BUILDING                              18
          AIRPORT BUILDING NON-TERMINAL - SECURE AREA      1
          AIRPORT TRANSPORTATION SYSTEM (ATS)              2
          AIRPORT/AIRCRAFT                                 4
          ALLEY                                          529
                                                        ... 
31.0      RESIDENTIAL YARD (FRONT/BACK)                    1
          SIDEWALK                                         2
          SMALL RETAIL STORE                               2
          STREET                                          12
          VEHICLE NON-COMMERCIAL                           2
Length: 2915, dtype: int64

In [41]:
# dropping the missing location description index row
chicago_crime_df.dropna(subset=['Location Description'],inplace=True)

In [42]:
chicago_crime_df.isna().sum()

Case Number             0
Date                    0
Block                   0
Primary Type            0
Location Description    0
Arrest                  0
Domestic                0
Beat                    0
District                0
Ward                    0
Community Area          0
Year                    0
Latitude                0
Longitude               0
dtype: int64

__3. convert inappropriate data types to the correct one__

We should convert the columns to appropriate datatypes for
1. Alignment with original dataset types.
2. For performance gains (clearly, using a numpy.float64 when a numpy.int64 is required will lower our memory footprint).

In [43]:
# create a copy of the dataframe to avoid conflict
c_crime_df = chicago_crime_df.copy()

In [44]:
# Including the date format dramatically increases performance.
c_crime_df["Date"] = pd.to_datetime(c_crime_df["Date"], format = "%m/%d/%Y %I:%M:%S %p", utc = True)
c_crime_df["Beat"] = c_crime_df["Beat"].astype(np.int64)
c_crime_df["District"] = c_crime_df["District"].astype(np.int64)
c_crime_df["Ward"] = c_crime_df["Ward"].astype(np.int64)
c_crime_df["Community Area"] = c_crime_df["Community Area"].astype(np.int64)
c_crime_df['Block'] = c_crime_df['Block'].astype(str)

# 4. Feature Engineering

Create new columns for Time, Date, Block_new, Street, Months, Day, and Season

In [45]:
# Create new columns for Months
def extract_and_map_month(date):
    # ensure the input is in datetime format
    # 07/26/2008 02:30:00 PM
    date_column = pd.to_datetime(date, format='%m/%d/%Y %I:%M:%S %p')

    month_names = date_column.dt.month_name()
# dt is an attribute of the date column object
# month_name() is a method
    return month_names

In [46]:

c_crime_df['Month'] = extract_and_map_month(c_crime_df['Date'])

# Group by month and count number of crimes
monthly_crime_counts = c_crime_df.groupby('Month').size()

In [47]:
# Create new columns for Day
def extract_and_map_month(date):
    # ensure the input is in datetime format
    # 07/26/2008 02:30:00 PM
    date_column = pd.to_datetime(date, format='%m/%d/%Y %I:%M:%S %p')

    day_names = date_column.dt.day_name()
# dt is an attribute of the date column object
# month_name() is a method
    return day_names

In [48]:
c_crime_df['Day'] = extract_and_map_month(c_crime_df['Date'])

# Group by day and count number of crimes
daily_crime_counts = c_crime_df.groupby('Day').size()

In [49]:
# Function to define various month and their corresponding season
def get_season(month):
    if month in ['December','January','February']:
        return 'Winter'
    elif month in ['March','April','May']:
        return 'Spring'
    elif month in ['June','July','August']:
        return 'Summer'
    else:
        return 'Fall'

In [50]:
c_crime_df['Season'] = c_crime_df['Month'].apply(get_season)

# Group by season and count number of crimes
seasonal_crime_counts = c_crime_df.groupby('Season').size()


In [51]:
c_crime_df.head()

Case Number                      Date                 Block  \
0    HP485721 2008-07-26 14:30:00+00:00  085XX S MUSKEGON AVE   
1    JB372830 2018-07-31 10:57:00+00:00     092XX S ELLIS AVE   
2    HN308568 2007-04-27 10:30:00+00:00     062XX N TRIPP AVE   
3    JB361368 2018-07-23 08:55:00+00:00    0000X N KEELER AVE   
4    JE297624 2021-07-11 06:40:00+00:00   016XX W HARRISON ST   

        Primary Type                    Location Description  Arrest  \
0    CRIMINAL DAMAGE                                  STREET   False   
1  WEAPONS VIOLATION                               POOL ROOM    True   
2           BURGLARY                               RESIDENCE    True   
3            ASSAULT            NURSING HOME/RETIREMENT HOME   False   
4            ASSAULT  PARKING LOT / GARAGE (NON RESIDENTIAL)   False   

   Domestic  Beat  District  Ward  Community Area  Year   Latitude  Longitude  \
0     False   423         4    10              46  2008  41.739980 -87.555120   
1     False   413         4     8              47  2018  41.726922 -87.599747   
2     False  1711        17    39              12  2007  41.994138 -87.734959   
3     False  1115        11    28              26  2018  41.881217 -87.730590   
4     False  1231        12    27              28  2021  41.874174 -87.668082   

   Month       Day  Season  
0   July  Saturday  Summer  
1   July   Tuesday  Summer  
2  April    Friday  Spring  
3   July    Monday  Summer  
4   July    Sunday  Summer

# 4. EDA: Data Exploration

## Non - Graphical

1. **Total Number of Crime cases Filed**

In [52]:
# Number of Unique Case Number.

print(f"Out of {len(c_crime_df)}: {len(c_crime_df['Case Number'].unique())} are unique case numbers")

Out of 2275836: 2275789 are unique case numbers


**Observation**: Most case numbers are unique case numbers.

Number of Non-Unique cases

In [53]:
# Out of all cases , how many are non_unique cases ?

non_unique_case_number = c_crime_df[c_crime_df.duplicated('Case Number') | c_crime_df.duplicated('Case Number', keep=False)]['Case Number'].tolist()

print(f"Length of non_unique: {len(non_unique_case_number)} out of {len(c_crime_df)}")

Length of non_unique: 92 out of 2275836


In [54]:
non_unique_case_number_df = c_crime_df[c_crime_df.duplicated('Case Number', keep=False)]
non_unique_case_number_df.head() # Saving non_unique case number dataframe for future use.

Case Number                      Date                 Block  \
25445     HM763064 2006-12-08 15:14:00+00:00    005XX W MADISON ST   
58104     JB483810 2018-10-21 02:39:00+00:00  001XX N LOCKWOOD AVE   
74485     HK172551 2004-02-08 23:44:00+00:00    034XX W MADISON ST   
80580     HV620483 2012-12-31 16:49:00+00:00       042XX W 21ST PL   
126290    HR146859 2009-01-31 01:10:00+00:00     039XX W ARGYLE ST   

       Primary Type Location Description  Arrest  Domestic  Beat  District  \
25445      HOMICIDE               OFFICE    True     False   111         1   
58104      HOMICIDE                 AUTO   False     False  1523        15   
74485      HOMICIDE               CHURCH   False     False  1123        11   
80580      HOMICIDE            APARTMENT    True      True  1012        10   
126290     HOMICIDE            APARTMENT    True     False  1712        17   

        Ward  Community Area  Year   Latitude  Longitude     Month       Day  \
25445     42              28  2006  41.881884 -87.640060  December    Friday   
58104     28              25  2018  41.883604 -87.757629   October    Sunday   
74485     28              27  2004  41.880909 -87.711843  February    Sunday   
80580     24              29  2012  41.852505 -87.730687  December    Monday   
126290    39              14  2009  41.971903 -87.726290   January  Saturday   

        Season  
25445   Winter  
58104     Fall  
74485   Winter  
80580   Winter  
126290  Winter

**Observation**: There are 2275789 unique values of case number out of 2275836 , it means there are 2275789 cases registered with a single crime the rest 92 are the case numbers which have commited multiple crimes.

2. **Out of total case registered how many are arrested and how many are not?**

In [55]:
# Out of total case registered how many are arrested and how many are not?


arrested = c_crime_df['Arrest'].where(c_crime_df['Arrest'] == True).count()
not_arrested = c_crime_df['Arrest'].where(c_crime_df['Arrest'] != True).count()

print(f"Out of {len(c_crime_df)},\n{arrested} are Arrested and \n{not_arrested} are not Arrested.")

Out of 2275836,
603465 are Arrested and 
1672371 are not Arrested.


**Observation**: Out of total crime cases received, 27% of the Criminals are arrested, 73% were not.

3. **How many unique crimes are available and what is the count for each crime from 2001-2022?**

In [56]:
# Group by Crime type and calculate count
crime_count = c_crime_df['Primary Type'].unique()
print(f'The Number of unique crimes are {len(crime_count)} and the count for each crime are:')
count_for_each_crime = pd.DataFrame(c_crime_df.groupby('Primary Type').size().sort_values(ascending=False).rename('Count').reset_index())
count_for_each_crime.head()

The Number of unique crimes are 35 and the count for each crime are:


Primary Type   Count
0            THEFT  479153
1          BATTERY  418647
2  CRIMINAL DAMAGE  259959
3        NARCOTICS  222861
4          ASSAULT  147662

4. **Number of crimes per Year**

In [57]:
# Count the number of crimes per year
crime_counts_per_year = c_crime_df['Year'].value_counts().sort_index().rename('Count').reset_index()
print(f'The amount of crimes per year are as follow: ')
crime_counts_per_year

The amount of crimes per year are as follow: 


Year   Count
0   2001  146292
1   2002  145441
2   2003  142708
3   2004  140802
4   2005  135737
5   2006  134769
6   2007  130754
7   2008  127883
8   2009  118376
9   2010  111077
10  2011  105639
11  2012  100724
12  2013   91802
13  2014   82597
14  2015   79183
15  2016   80872
16  2017   80312
17  2018   80702
18  2019   77719
19  2020   63193
20  2021   62284
21  2022   36970

5. **Number of crimes by month**

how many records we have for each month ?

In [58]:
# Count the number of crimes per Month
crime_counts_per_month = c_crime_df['Month'].value_counts().sort_index().rename('Count').reset_index()
print(f'The amount of crimes per month are as follow: ')
crime_counts_per_month

The amount of crimes per month are as follow: 


Month   Count
0       April  187239
1      August  206089
2    December  167906
3    February  158349
4     January  179894
5        July  213452
6        June  204809
7       March  188532
8         May  204425
9    November  176376
10    October  195487
11  September  193278

6. **Number of crimes by seasons**

how many records we have for each seasons ?

In [59]:
# Count the number of crimes per Seasons
crime_counts_per_season = c_crime_df['Season'].value_counts().sort_index().rename('Count').reset_index()
print(f'The Seasonal Crime rate are as follow: ')
crime_counts_per_season

The Seasonal Crime rate are as follow: 


Season   Count
0    Fall  565141
1  Spring  580196
2  Summer  624350
3  Winter  506149

7. **Number of crimes by day**

how many records we have for each day ?

In [60]:
# Count the number of crimes per day
crime_counts_per_day = c_crime_df['Day'].value_counts().sort_index().rename('Count').reset_index()
print(f'The daily Crime rate are as follow: ')
crime_counts_per_day

The daily Crime rate are as follow: 


Day   Count
0     Friday  342109
1     Monday  321298
2   Saturday  326139
3     Sunday  309264
4   Thursday  323885
5    Tuesday  325467
6  Wednesday  327674

In [61]:
# What are the top 5 most common crime types?
common_crime = c_crime_df['Primary Type'].value_counts().head()
print(f'The top 5 most common crime types are {common_crime}')

The top 5 most common crime types are Primary Type
THEFT              479153
BATTERY            418647
CRIMINAL DAMAGE    259959
NARCOTICS          222861
ASSAULT            147662
Name: count, dtype: int64


In [62]:
# what is the most common location description for crimes?
print("Most common location description for crimes:",c_crime_df['Location Description'].mode()[0])

Most common location description for crimes: STREET


In [63]:
# what are the top 10 most common crimes in 2021?
crime_2021 = c_crime_df[c_crime_df['Year'] == 2021]
print("Top 10 most crime in 2021: ")
crime_2021['Primary Type'].value_counts().head(10)

Top 10 most crime in 2021: 


Primary Type
THEFT                  12313
BATTERY                12017
CRIMINAL DAMAGE         7571
ASSAULT                 6134
DECEPTIVE PRACTICE      4905
OTHER OFFENSE           4120
MOTOR VEHICLE THEFT     3163
WEAPONS VIOLATION       2752
ROBBERY                 2407
BURGLARY                1978
Name: count, dtype: int64

In [64]:
# How many crimes were reported each month in 2021 ?
monthly_crime_2021 = crime_2021['Month'].value_counts().sort_index()
print('Number of crimes reported each month in 2021:')
monthly_crime_2021

Number of crimes reported each month in 2021:


Month
April        4609
August       5590
December     5197
February     3851
January      4848
July         5630
June         5585
March        4839
May          5478
November     5127
October      5809
September    5721
Name: count, dtype: int64

8. **Top 10 - Number of cases each day**

In [ ]:
# Top 10 - Number of cases each day?

top_10_num_of_cases_each_day = c_crime_df.groupby(['Date'])['Case Number'].count().reset_index(name = 'no_of_cases').set_index('no_of_cases').sort_index(ascending = False)[:10]
top_10_num_of_cases_each_day

## Graphical

In [ ]:
crime_counts_per_year = c_crime_df['Year'].value_counts().sort_index()
# Plotting
plt.figure(figsize=(20, 13))
crime_counts_per_year.plot(kind='bar',fontsize = 15)
plt.title('Trend of Crimes in Chicago from 2001 to 2022', fontsize = 30)
plt.xlabel('Year',fontsize = 20)
plt.ylabel('Number of Crimes',fontsize = 20)
plt.xticks(rotation=45)
plt.grid(axis='y', linestyle='--', alpha=0.7)
plt.tight_layout()
plt.show()

**Observations:**

1. The crime rate is dropping overall.
2. In 2001,number of crime incidents were close to 150k.
3. In 2022 ,it's estimate to be 36k , so it's a major drop in crime.
4. By 2023 we expect a major drop down to an estimate of 10k

5. **Number of crimes by month**

how many records we have for each month ?





In [ ]:
crime_counts_per_month = c_crime_df['Month'].value_counts().sort_index()
# Plotting
plt.figure(figsize=(20, 13))
crime_counts_per_month.plot(kind='bar',fontsize = 20)
plt.title('Monthly Trend of Crimes in Chicago', fontsize = 30)
plt.xlabel('Months',fontsize = 20)
plt.ylabel('Number of Crimes',fontsize = 20)
plt.xticks(rotation=45)
plt.grid(axis='y', linestyle='--', alpha=0.7)
plt.tight_layout()
plt.show()

In [ ]:

# monthly_crime_count = c_crime_df.resample("M"  , on = 'Date').size()

# # Create a line plot.

# plt.figure(figsize = (10,6))
# plt.plot(monthly_crime_count.index , monthly_crime_count.values , marker = 'o')
# plt.xlabel("Date")
# plt.ylabel("Number of crime incidents")
# plt.title("Monthly Crime Trends in Chicago")
# plt.xticks(rotation = 45)
# plt.tight_layout()
# plt.show()


This chart clearly shows a "periodic" pattern in the crimes over many years, which shows why crimes are very predictable activity.

**Observations:**
- From the figure above, its observed that FEBRUARY recorded the least amount of crime cases and JULY is the month where crimes are comitted the most.

6. **Number of crimes by Season**

how many records we have for each season ?

In [ ]:
crime_counts_per_season = c_crime_df['Season'].value_counts().sort_index()
# Plotting
plt.figure(figsize=(20, 13))
crime_counts_per_season.plot(kind='bar',fontsize = 25)
plt.title('Seasonal Trend of Crimes in Chicago', fontsize = 30)
plt.xlabel('Season',fontsize = 20)
plt.ylabel('Number of Crimes',fontsize = 20)
plt.xticks(rotation=45)
plt.grid(axis='y', linestyle='--', alpha=0.7)
plt.tight_layout()
plt.show()

In [ ]:
crime_counts_per_day = c_crime_df['Day'].value_counts().sort_index()
# Plotting
plt.figure(figsize=(20, 13))
crime_counts_per_day.plot(kind='bar',fontsize = 25)
plt.title('Daily Trend of Crimes in Chicago', fontsize = 30)
plt.xlabel('Day',fontsize = 20)
plt.ylabel('Number of Crimes',fontsize = 20)
plt.xticks(rotation=45)
plt.grid(axis='y', linestyle='--', alpha=0.7)
plt.tight_layout()
plt.show()

In [ ]:
# Top 10 - Number of cases each day
top_10_num_of_cases_each_day_df = pd.DataFrame(top_10_num_of_cases_each_day)
index_df = top_10_num_of_cases_each_day_df.index

sns.barplot(x = index_df , y = 'Date' , data = top_10_num_of_cases_each_day_df)
plt.title("Top 10 - Number of cases each day - Bar plot")
plt.xlabel("Number of Cases")
plt.ylabel("Date")
plt.show()

**Observation**: The average Number of cases per day could be less than 60


7. __Number of crime by type__</br>
We try to indentify the trend and highpoints or lowpoints.

In [ ]:
# Function to plot the serie value_counts
def plot_counts(serie, title):
    df = pd.DataFrame(serie.value_counts()[:10])
    df.columns = ["Freq"]
    df["Type"] = df.index
    fig = px.bar(df, y="Freq", x="Type", text="Freq", color="Freq", color_continuous_scale=px.colors.sequential.Blugrn)
    fig.update_traces(texttemplate="%{text:.2s}", textposition="outside")
    fig.update_layout(uniformtext_minsize=8, uniformtext_mode="hide")
    fig.update_layout(title_text=title)
    fig.show()


plot_counts(c_crime_df["Primary Type"], "Type of Crimes")

Observation: THEFT, BATTERY, CRIMINAL DAMAGE and NARCOTICS represents more the 65% of the crimes.

8. __Where does crimes happens the most?__

In [ ]:
plot_counts(c_crime_df["Location Description"], "Location of Crimes")

Crime happens mostly on the street

9. __Number of crimes per Year x Months__ </br>
Let's see the number of crimes per Year and Months...

In [ ]:
# first we convert the date column to datetime data type.

c_crime_df['Date'] = pd.to_datetime(chicago_crime_df['Date'],format = "%m/%d/%Y %I:%M:%S %p")
c_crime_df.set_index("Date", inplace=True)

In [ ]:
cases = pd.DataFrame(c_crime_df.resample("M").size(), columns=["Number of cases"])
cases["Month"] = cases.index.month.astype(str)
cases["Year"] = cases.index.year.astype(str)
cases["Year-Month"] = cases["Year"].str.cat(cases["Month"].str.zfill(2), sep="-")
cases.head()

BY HEATMAP

In [ ]:
case_table = cases[["Month", "Year", "Number of cases"]]
case_table  = case_table.pivot(index='Month', columns='Year', values='Number of cases').fillna(0).astype(int)
case_table.index = case_table.index.astype(int)
case_table = case_table.sort_index()
case_table.index = ['Jan', 'Fev', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec']
plt.figure(figsize=(20, 10))
_ = sns.heatmap(
    case_table,
    annot=True,
    fmt="d",
    linewidths=0.5,
    cmap="Blues",
).set_title('Number of crimes in chicago per Year x Months')

The years between 2001 and 2012 were a tough time, the year 2008 was the worst with a mean of 71004 crimes per month, and the best year was the 2015 with a mean of 21916 crimes per month.

BY BAR

In [ ]:
fig = px.bar(
    cases,
    x="Year-Month",
    y="Number of cases",
    hover_data=["Year", "Month", "Number of cases"],
    color="Number of cases",
    height=600,
    width=1400,
    color_continuous_scale=px.colors.sequential.Blugrn

)
fig.update_layout(title_text='Crime counts per year-month in bars')
fig.show()

Similar graph but using bars to visualize though noisy but can be used. Again the 'tough time' during 2001 until 2012.

10. **Are there specific location that crimes used to happen?**

Using Map plot

In [ ]:
map_marks = chicago_crime_df[['Latitude', 'Longitude']]
map_marks['Year'] =  pd.to_datetime(chicago_crime_df.Date, format="%m/%d/%Y %I:%M:%S %p").dt.year
map_marks.head()

In [ ]:
fig = px.scatter_mapbox(map_marks[map_marks.Year == 2016].sample(5000), lat="Latitude", lon="Longitude",
                        color_discrete_sequence=["fuchsia"], zoom=9, height=650)
fig.update_layout(mapbox_style="open-street-map")
fig.update_layout(margin={"r":0,"t":0,"l":0,"b":0})
fig.show()

Look like there is crime everywhere in Chicago, in the centre, you can see more points but also is a crowded place.

11. **What is the arrested rate per crime?**

In [ ]:
cases = c_crime_df.groupby('Primary Type')['Arrest'].mean().reset_index()
cases['Arrest (%)'] = cases['Arrest'].round(4)*100
cases.sort_values('Arrest', inplace=True)
fig = px.bar(
    cases,
    orientation='h',
    x="Arrest (%)",
    y="Primary Type",
    color="Arrest (%)",
    text="Arrest (%)",
    height=800,
    #width=800,
        color_continuous_scale=px.colors.sequential.Blugrn

)
fig.update_traces(texttemplate="%{text:.4s}", textposition="outside")
fig.update_layout(title_text="Arrested Rate in % per Type Crime in Chicago.")

fig.show()

The “VIOLATION” crimes look strongly correlated to the arrested rate once the 'cop' must be present in the location to declare the violation.

12. **Top 10 Block with the most crime**

In [ ]:
# Group the data by 'Block' and count occurrences
block_counts = chicago_crime_df['Block'].value_counts()

# Select the top 10 blocks with the most crimes
top_10_blocks = block_counts.head(10)

# Plot the top 10 blocks
top_10_blocks.plot(kind='bar', figsize = (20,13), fontsize = 15)
plt.title('Top 10 Blocks with Most Crime in Chicago',fontsize=30)
plt.xlabel('Block', fontsize=20)
plt.ylabel('Number of Crimes',fontsize=20)
plt.xticks(rotation=90, ha='right')
plt.tight_layout()
plt.show()

13. **Top - 10 Ward with most amount of crimes**

In [ ]:
# Group by 'Ward' and count the occurrences of crimes in each ward
ward_crime_counts = chicago_crime_df['Ward'].value_counts()

# Select the top 10 wards
top_10_wards = ward_crime_counts.head(10)

# Plot the data
top_10_wards.plot(kind='bar', figsize = (20,13), fontsize = 15)
plt.title('Top 10 Wards with the Most Amount of Crimes', fontsize = 30)
plt.xlabel('Ward', fontsize = 20)
plt.ylabel('Number of Crimes', fontsize = 20)
plt.xticks(rotation=45)
plt.tight_layout()
plt.show()

14. **Top - 10 community area with the most amount of crimes**




In [ ]:
# Group the data by community area and count the number of crimes
community_crime_count = c_crime_df['Community Area'].value_counts().reset_index()
community_crime_count.columns = ['Community Area', 'Crime Count']

# Sort the data to find the top 10 community areas with the most crimes
top_10_community_crime = community_crime_count.head(10)

# Visualize the top 10 community areas with the most crimes
plt.figure(figsize=(20, 13))
sns.barplot(x='Community Area', y='Crime Count', data=top_10_community_crime, palette='viridis')
plt.title('Top 10 Community Areas with the Most Crimes in Chicago', fontsize = 30)
plt.xlabel('Community Area',fontsize = 20)
plt.ylabel('Number of Crimes', fontsize = 20)
plt.xticks(rotation=45)
plt.tight_layout()
plt.show()

15. **Top - 10 communities with the most crime in 2023**

In [ ]:
# Filter the dataset for the year 2023
chicago_crime_2023 = c_crime_df[c_crime_df['Year'] == 2023]

# Group by community and count the number of crimes
crime_by_community = chicago_crime_2023['Community Area'].value_counts().sort_values(ascending=False)

# Select the top 10 communities with the most crime
top_10_communities = crime_by_community.head(10)

# Plot the top 10 communities with the most crime
plt.figure(figsize=(20, 13))
top_10_communities.plot(kind='bar', fontsize = 15)
plt.title('Top 10 Communities with the Most Crime in 2023', fontsize = 30)
plt.xlabel('Community Area',fontsize = 20)
plt.ylabel('Number of Crimes',fontsize = 20)
plt.xticks(rotation=45)
plt.tight_layout()
plt.show()

16. **Top - 10 Location  where theft normally occurs**

In [ ]:
# Filter the dataset to include only theft-related incidents
theft_df = chicago_crime_df[chicago_crime_df['Primary Type'] == 'THEFT']

# Count the occurrences of theft incidents at each location
theft_locations_count = theft_df['Location Description'].value_counts()

# Get the top 10 locations
top_10_theft_locations = theft_locations_count.head(10)

# Plot the top 10 theft locations
plt.figure(figsize=(20, 13))
top_10_theft_locations.plot(kind='bar',fontsize = 15)
plt.title('Top 10 Locations for Theft in Chicago',fontsize = 30)
plt.xlabel('Location',fontsize = 20)
plt.ylabel('Number of Incidents', fontsize = 20)
plt.xticks(rotation=45, ha='right')
plt.tight_layout()
plt.show()

17. **Top-15 crime where arrest occured**

In [ ]:
ax = c_crime_df[c_crime_df['Arrest'] == True].value_counts("Primary Type")[:15]\
    .plot(kind = 'barh',
          figsize = (20,13),
          fontsize = 30)
ax.set_ylabel("Type of Crime", fontsize=20)
ax.set_xlabel("Count", fontsize=20)
ax.set_title("Top 15 crimes were arrests occured", fontsize=30)
plt.show()
